### ISSUE TITLE
Supporting multiple concentrations and pre-defined epitopes for EscapeModel

### ISSUE THREAD

**Issue by timcyu** [matsengrp/torchdms#120]
## Learning data measured at multiple antibody mixture concentrations

We are interested in measuring escape fractions of mutants at multiple concentrations. This will add an additional column in the dataset, like so:

![Screen Shot 2021-07-09 at 11 32 04 AM](https://user-images.githubusercontent.com/36126324/125125902-246db400-e0af-11eb-942d-d33e0a99530b.png)

We would like for the concentration to be included in the model fitting procedure. Therefore, we will need to:
1. add a way to read in the concentration (so that for each mutant, we know what concentration its escape fraction was measured at).
2. modify the sigmoid transformation to multiply the exponentiated term by `c`, the concentration. Currently, the code for going from the latent representation -> escape fraction looks like:
```python
def from_latent_to_output(self, x): 
        """latent space in as 'x' -> escape fraction."""
        b_fractions = torch.sigmoid(x)
        return torch.unsqueeze(torch.prod(b_fractions, 1), 1)
``` 
I'm imagining that we can modify this with a custom sigmoid function that looks something like:
```python
# PSEUDOCODE
def from_latent_to_output(self, x, c): 
        """latent space in as 'x' -> escape fraction."""
        b_fractions = 1 / (1 + c * torch.exp(-x))
        return torch.unsqueeze(torch.prod(b_fractions, 1), 1)
``` 

## Pre-defining epitopes

The goal of pre-defining epitopes is to use the results of prior structural biology studies to ensure that each latent phenotype (or hidden layer nodes) in the model can be interpreted as an **independent** epitope. To do this, we will need:
1. A data object that encodes the sites that are within each epitope. This can be a binary tensor of size `(E,S)`. If site `s` in `S` is located in site `e` in `E`, then 1, otherwise 0.
2. **Epitope mask**: A way to apply the data object on the input data prior to entering the hidden layer. This way, any mutations at sites that are not present in the epitope will be multiplied by 0, but all mutations at sites within the epitope will be multiplied by 1. Currently, the code for going from the input data -> latent representation looks like:
```python
def to_latent(self, x):
        """input features -> latent space."""
        latent_dims = []
        for i in range(self.num_epitopes):
            model_ = getattr(self, f"latent_layer_epi{i}")
            latent_dims.append(model_(x))

        return torch.cat(latent_dims, dim=1)
```
I couldn't fully recover my old code for applying the epitope mask, but assuming that `epitope_weights` is an `(E,S)` tensor, I think we can modify the function like this:
```python
# PSEUDOCODE
def to_latent(self, x):
        """input features * epitope_weights -> latent space."""
        # expand weights to account for all aa alphabet letters
        expanded_weights = torch.repeat_interleave(self.epitope_weights, 21, dim=1)

        latent_dims = []
        for i in range(self.num_epitopes):
            input_ = x * expanded_weights[i]
            model_ = getattr(self, f"latent_layer_epi{i}")
            latent_dims.append(model_(input_))

        return torch.cat(latent_dims, dim=1)
```
